### PR TITLE
Set NotNull constraints for invocations in SE

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Contributes to #5860

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 1]
```diff
@@ -567,5 +569,97 @@ public void Main<T>() where T : new()
...
+}
+
+public static class Extensions
+{
+    public static void ExtensionMethod(this Sample s) { }
+    public static void ExtensionMethod(this bool b) { }
+}";
+            var validator = new SETestContext(code, AnalyzerLanguage.CSharp, Array.Empty<SymbolicCheck>()).Validator;
+            validator.ValidateContainsOperation(OperationKind.Invocation);
+            validator.ValidateTag("BeforeInstance", x => x.Should().BeNull());
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:609]: We have tons of validations on `SymbolicValue`, `SymbolicContext` and other related types. We should consider adding custom validations to increase readability and remove duplication. Maybe a small suite of assertions like this:
```cs
validator.ValidateTag("BeforeInstance", x => x.Should().HaveNoConstraints());
validator.ValidateTag("AfterInstance", x => x.Should().HaveConstraint(ObjectConstraint.NotNull, "Instance method should set NotNull constraint."));
validator.ValidateTag("AfterInstance", x => x.Should().HaveOnlyConstraint(ObjectConstraint.NotNull, "Instance method should set NotNull constraint."));
validator.ValidateTag("AfterInstance", x => x.Should().HaveAtLeastConstraint(ObjectConstraint.NotNull, SomeOtherConstraint.Value);
```
See [fluentassertions extensibility](https://fluentassertions.com/extensibility/). A sample can also be found in #5880: https://github.com/SonarSource/sonar-dotnet/blob/a05abc4c2cddee7823e85b35307f27f788c69312/analyzers/tests/SonarAnalyzer.UnitTest/Extensions/BlockSyntaxExtensionsTests.cs#L273-L314

What do you think?

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM. One optional test could be added, though.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs [hunk 2]
```diff
@@ -567,5 +569,97 @@ public void Main<T>() where T : new()
...
+            validator.ValidateTag("BeforeExtension", x => x.Should().BeNull());
+            validator.ValidateTag("BeforePreserve", x => x.HasConstraint(BoolConstraint.True).Should().BeTrue());
+            validator.ValidateTag("AfterInstance", x => x.HasConstraint(ObjectConstraint.NotNull).Should().BeTrue("Instance method should set NotNull constraint."));
+            validator.ValidateTag("AfterExtension", x => x.Should().BeNull("Extensions can run on null instances."));
```
[the remark this answers is not in the document]
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:613]: It's more a concern for the rule than the engine, but I can add a test case

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Operations.cs:609]: We can consider that. It should look like
```
validator.TagValue("BeforeInstance").Should().HaveNoConstraints();
```

**Pull request merged**